### PR TITLE
Spin out driver subclass

### PR DESCRIPTION
**Pull request by zsyed91**
- Create a `Driver` subclass
- Added more tests to consider some edge cases and sad paths
- Test coverage is 100%
- Use `FileWatcher` instead of `IO.select` to prevent 100% cpu usage when waiting for value to change. Method now blocks until value is updated.

### CODE REVIEW

File: lib/pi_piper/sysfs.rb [hunk 1]
```diff
@@ -1,6 +1,7 @@
-require "pi_piper/sysfs/version"
-require "pi_piper/sysfs/sysfs"
+require 'pi_piper'
+require 'pi_piper/sysfs/version'
+require 'pi_piper/sysfs/driver'
 
 module PiPiper
-  self.driver= PiPiper::Sysfs
+  self.driver = PiPiper::Sysfs::Driver.new
```
Comment (elmatou) [inline lib/pi_piper/sysfs.rb]: Unless you changed PiPiper's code `#driver=` need a subclass of Driver, not an instance.

File: lib/pi_piper/sysfs/driver.rb [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+      end
+
+      private
+
+      def export(pin)
+        raise RuntimeError, "pin #{pin} is already reserved by another Pin instance" if @exported_pins.include?(pin)
+        File.write("/sys/class/gpio/export", pin)
+        @exported_pins << pin
+      end
+
```
Comment (elmatou) [inline lib/pi_piper/sysfs/driver.rb:87]: following methods are not useful.
I removed them once, because, they only are used one time (maybe twice for the value_file) but doesn't make code cleaner or easier to understand.

File: lib/pi_piper/sysfs/driver.rb [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+      def pin_set_trigger(pin, trigger)
+        raise ArgumentError, "trigger should be :falling, :rising, :both or :none" unless [:falling, :rising, :both, :none].include? trigger
+        raise ArgumentError, "Pin #{pin} not exported" unless exported?(pin)
+        File.write(edge_file(pin), trigger)
+      end
+
+      def pin_wait_for(pin, trigger)
+        pin_set_trigger(pin, trigger)
+        value = pin_read(pin)
+
```
Comment (elmatou) [inline lib/pi_piper/sysfs/driver.rb:50]: I was not aware of any CPU usage issue with the previous implementation, it was backed by a unix system call `select()` which is widely used since august 1984... and worked pretty well since I used PiPiper also (maybe a year).
can you please explain me a bit more the issue ?

on your implementation : This need to be discussed ; your are watching the value_file and not the edge_file, why ? Is it wise to add a new dependecy to the gem ? FileWatcher seems to use File#mtime to check if a file is changed, is it reliable ? nothing in the gpio kernel module is said about dates of the value or edge files.

Let's talk.

EDIT : My bad select(2) must watch the value_file obviously !

File: lib/pi_piper/sysfs/driver.rb (hunk 2 above)
Comment (elmatou) [inline lib/pi_piper/sysfs/driver.rb:87]: But I won't make a case for it !

File: pi_piper-sysfs.gemspec [hunk 4]
```diff
@@ -24,6 +24,7 @@ Gem::Specification.new do |spec|
   spec.require_paths = ["lib/pi_piper"]
 
   spec.add_runtime_dependency 'pi_piper', ">= 2.0.0"
+  spec.add_runtime_dependency 'filewatcher', '~> 0.5.3'
```
Comment (elmatou) [inline pi_piper-sysfs.gemspec]: see above

File: lib/pi_piper/sysfs/driver.rb [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+      def edge_file(pin)
+        "/sys/class/gpio/gpio#{pin}/edge"
+      end
+
+      def direction_file(pin)
+        "/sys/class/gpio/gpio#{pin}/direction"
+      end
+
+      def pin_value_changed?(pin, trigger, value)
```
Comment (elmatou) [inline lib/pi_piper/sysfs/driver.rb:100]: see above

File: spec/pi_piper/sysfs/driver_spec.rb [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+    it 'should return true if pin is exported' do
+      subject.pin_direction(4, :in)
+      expect(subject.exported?(4)).to be(true)
+    end
+
+    it 'should return false if pin is not exported' do
+      expect(subject.exported?(4)).to be(false)
+    end
+  end
+end
```
Comment (elmatou) [inline spec/pi_piper/sysfs/driver_spec.rb:211]: You did a great work by expanding the specs !! you'll have to explain to me why you absolutely want a 100% coverage (as count he gem) ? I'm pretty sure the previous specs did detect the same edge cases.

**Comment by elmatou** [comment]
Hi @zsyed91, 
My thoughts lies in the comments above.
I'm only concerned on your `#wait_for_change` implementation, maybe you can explained me what pushed you to this direction.

File: lib/pi_piper/sysfs.rb (hunk 1 above)
Comment (zsyed91) [inline lib/pi_piper/sysfs.rb]: Yes you are correct, forgot to remove the `.new`. I will fix this.

File: lib/pi_piper/sysfs/driver.rb (hunk 2 above)
Comment (zsyed91) [inline lib/pi_piper/sysfs/driver.rb:87]: Sure I can see your point. I was just thinking along 2 points. Anything called twice or more should be in a method. And second, we could stub out the method to test another file when writing tests.

File: lib/pi_piper/sysfs/driver.rb (hunk 3 above)
Comment (zsyed91) [inline lib/pi_piper/sysfs/driver.rb:50]: Sure. I thought the original implementation watched the value file and compared whether the value changed or not? And the reason I implemented it this way is because the original implementation actually spins at 100%. I verified this by running the original method manually in an `irb` session. I can double check this again but that was my inspiration. I agree about adding another gem, I wasn't too keen on it but I had difficulty finding some sample code that would block the script until a file write. This gem apparently does it and seems to work well. We should definitely run it through some stuff to verify performance though.

But basically that is what my thinking was. It actually doesn't use File#mtime but rather listens for kernel events. Having said that I did not look at the source code in great detail but the readme mentioned that part. If that is the case then we ultimately hit the kernel anyway so this method may be more robust. I know every OS has a "file system watcher" event handler so hopefully this one is using one in the code.

File: spec/pi_piper/sysfs/driver_spec.rb (hunk 6 above)
Comment (zsyed91) [inline spec/pi_piper/sysfs/driver_spec.rb:211]: sure, we don't necessarily need 100% coverage but why not :smile:

**Comment by zsyed91** [comment]
@elmatou see my note in the code regarding `FileWatcher`. We can definitely discuss it further, it is a big change, after all.

File: lib/pi_piper/sysfs/driver.rb (hunk 3 above)
Comment (zsyed91) [inline lib/pi_piper/sysfs/driver.rb:50]: Here are some stackoverflow questions I was researching when testing the original implementation:

http://stackoverflow.com/questions/11367719/why-is-my-ruby-script-utilizing-90-of-my-cpu
http://linux.die.net/man/2/select
(there are a couple more, I will have to re-search them on google but will link them once I get home)

Apparently `select` does not block because the `read` does not block when at `EOF` but rather returns `nil`. This is why it consumes the CPU. If you look at the man page for select it even states that it will only block if the read blocks, this is not the case for us. Read will always succeed.

Comment (elmatou) [inline lib/pi_piper/sysfs/driver.rb:50]: Hi @zsyed91,
I tried to produce what you were speeking about. see the gist I wrote to do it :
https://gist.github.com/elmatou/4fd0edb39ea278c465b211ddbb773f45
While I was watching the cpu usage in another ssh session :
`grep 'cpu ' /proc/stat | awk '{usage=($2+$4)*100/($2+$4+$5)} END {print usage "%"}'`

It happens I always get something around 11.4% more or less....

Of course I'm looking at the edge file, which is not the same thing the value file which cannot have an EOF (it is updated on every clock round).

Comment (elmatou) [inline lib/pi_piper/sysfs/driver.rb:50]: note that @jwhitehorn implementation look at the value_file as mine look at the edge_file, with which one did you hit 100% ?

**Comment by elmatou** [comment]
Hi @zsyed91,
tell me when your are good to go with your PR.
I just succeed to use I2C with sysfs & bcm2835. This will be a nice addition. see you

**Comment by zsyed91** [comment]
@elmatou Yeah I have been insanely busy this past week with some life stuff. I will push the suggestions soon. Glad you tested I2C with the code and that it works :smile:

**Comment by elmatou** [comment]
My bad, select was and should listen to value_file of course, edge_file is only for setup.

As far I understand, FileWatcher use modification time, I get through the code and only found this line referring to the file :
https://github.com/thomasfl/filewatcher/blob/master/lib/filewatcher.rb#L125

as for select(2) not blocking on read. We need to investigate more, the SO thread you mentioned was listening on the reading part of the file descriptor and not the error part hence the difference in behaviour. 
looking forward to read you.

**Comment by zsyed91** [comment]
Sure I can take a look again. I will have to look at the difference between checking errors vs just reading via select. If it helps, I can pull that logic out of this PR and open a separate one containing those changes.

**Comment by elmatou** [comment]
I think it would be best to keep atomic/feature oriented PR. I will
continue to investigate on the matter.

**Comment by elmatou** [comment]
Hi @zsyed91, 
where are we on this since we last spoke ?
do you need a hand to spin out the drivers ?

**Comment by zsyed91** [comment]
@elmatou Hey, sorry I just switched jobs and am currently moving cross-country so this slipped through the cracks. I just pushed the changes. I fixed the driver class and removed the `pin_wait_for` changes so that I can open it in a separate PR.

**Comment by elmatou** [comment]
Ok, thx you, I'm gonna do a full review of the PR, and it will be good to merge

**Comment by zsyed91** [comment]
Awesome, let me know and then I will merge if everything is good to go :smile:

File: spec/spec_helper.rb [hunk 7]
```diff
@@ -1,3 +1,9 @@
+module PiPiper
+  def self.driver=(driver)
+    # Placeholder for testing, PiPiper 3.0 contains this so we can remove
+  end
```
Comment (elmatou) [inline spec/spec_helper.rb:4]: you are right, we should remove this.

File: lib/pi_piper/sysfs/driver.rb (hunk 5 above)
Comment (elmatou) [inline lib/pi_piper/sysfs/driver.rb:100]: where do you use this method, I can't see. maybe we can halt on this until we sort the #pin_wait_for issue

**Comment by elmatou** [comment]
I just pointed to 2 methods not useful for now.
and maybe we could set the dependency to pi_piper 3.0 (if 3.0 is good for you.).

File: spec/spec_helper.rb (hunk 7 above)
Comment (zsyed91) [inline spec/spec_helper.rb:4]: right, once we release 3.0 officially I can open a new pr to remove this.

File: lib/pi_piper/sysfs/driver.rb (hunk 5 above)
Comment (zsyed91) [inline lib/pi_piper/sysfs/driver.rb:100]: Ah this was actually leftover from my implementation of `#pin_wait_for`. I will go ahead and remove this for now and add it to the other PR.
